publish-mqtt/src/homie.rs

```diff
@@ -100,6 +105,78 @@ enum State {
     Ready,
 }
 
+/// Builder for `HomieDevice` and associated objects.
+pub struct HomieDeviceBuilder {
+    device_base: String,
+    device_name: String,
+    firmware_name: String,
+    firmware_version: String,
+    mqtt_options: MqttOptions,
+}
+
+impl HomieDeviceBuilder {
+    /// Set the firmware name and version to be advertised for the Homie device.
+    ///
+    /// If this is not set, it will default to the cargo package name and version.
+    #[allow(dead_code)]
+    pub fn set_firmware(&mut self, firmware_name: &str, firmware_version: &str) {
+        self.firmware_name = firmware_name.to_string();
+        self.firmware_version = firmware_version.to_string();
+    }
+
+    /// Create a new Homie device, connect to the MQTT server, and start a task to handle the MQTT
+    /// connection.
+    ///
+    /// # Return value
+    /// A pair of the `HomieDevice` itself, and a `Future` for the tasks which handle the MQTT
+    /// connection. You should join on this future to handle any errors it returns.
+    pub async fn spawn(
+        mut self,
+    ) -> Result<
+        (
+            HomieDevice,
+            impl Future<Output = Result<(), Box<dyn Error + Send + Sync>>>,
+        ),
+        SendError<Request>,
+    > {
+        self.mqtt_options.set_last_will(LastWill {
+            topic: format!("{}/$state", self.device_base),
+            message: "lost".to_string(),
+            qos: QoS::AtLeastOnce,
+            retain: true,
+        });
+        let mut event_loop = EventLoop::new(self.mqtt_options, REQUESTS_CAP).await;
+
+        let mut homie = HomieDevice::new(
+            event_loop.handle(),
+            self.device_base,
+            self.device_name,
+            self.firmware_name,
+            self.firmware_version,
+        );
+
+        let stats = HomieStats::new(event_loop.handle(), homie.device_base.to_string());
+
+        // This needs to be spawned before we wait for anything to be sent, as the start() calls below do.
+        let event_task: JoinHandle<Result<(), Box<dyn Error + Send + Sync>>> =
+            task::spawn(async move {
+                loop {
+                    let (incoming, outgoing) = event_loop.poll().await?;
+                    log::trace!("Incoming = {:?}, Outgoing = {:?}", incoming, outgoing);
+                }
+            });
+
+        stats.start().await?;
+        homie.start().await?;
+
+        let stats_task: JoinHandle<Result<(), Box<dyn Error + Send + Sync>>> =
+            task::spawn(stats.run());
```

Comment:
I *think* that you can skip this type annotation because the type is fully specified elsewhere already (no ? in sight here).

Reply:
I'll fix this in the next refactor I'm working on.